api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,758 @@
+		return err
+	}
+
+	humanReadableProviders := ""
+	for _, r := range resourceProviders {
+		humanReadableProviders += fmt.Sprintf("\n    %#v", r)
+	}
+
+	log.Printf("[DEBUG] [EnvironmentScaler] resourceProviders for env '%s': %s", environmentID, humanReadableProviders)
+
+	var resourceConsumers []ResourceConsumer
+
+	// get pending service resource consumers
+	serviceConsumers, err := e.getResourceConsumers_PendingServices(clusterName)
+	if err != nil {
+		return err
+	}
+
+	resourceConsumers = append(resourceConsumers, serviceConsumers...)
+
+	// get pending task resource consumers in ECS
+	ecsTaskConsumers, err := e.getResourceConsumers_TasksInECS(clusterName)
+	if err != nil {
+		return err
+	}
+
+	resourceConsumers = append(resourceConsumers, ecsTaskConsumers...)
+
+	// get pending task resource consumers in jobs
+	jobTaskConsumers, err := e.getResourceConsumers_TasksInJobs(clusterName)
+	if err != nil {
+		return err
+	}
+
+	resourceConsumers = append(resourceConsumers, jobTaskConsumers...)
+
+	log.Printf("[DEBUG] [EnvironmentScaler] resourceConsumers for env '%s': %#v", environmentID, resourceConsumers)
+
+	var errs []error
+
+	// calculate for scaling up
+	resourceProviders, scaleUpErrs := e.calculateScaleUp(clusterName, resourceProviders, resourceConsumers)
+
+	errs = append(errs, scaleUpErrs...)
+
+	// calculate for scaling down
+	unusedProviders := e.calculateScaleDown(clusterName, resourceProviders)
+
+	// do the scaling
+	desiredScale := len(resourceProviders) - len(unusedProviders)
+
+	if err := e.scaleTo(clusterName, desiredScale, unusedProviders); err != nil {
+		errs = append(errs, err)
+	}
+
+	return errors.MultiError(errs)
+}
+
+func (e *EnvironmentScaler) calculateNewProvider(clusterName string) (*ResourceProvider, error) {
+	env, err := e.EnvironmentProvider.Read(clusterName)
+	if err != nil {
+		return nil, err
+	}
+
+	resource := &ResourceProvider{}
+	resource.AvailableMemory = InstanceSizes[env.InstanceSize]
+	resource.ID = "<new instance>"
+	resource.InUse = false
+	resource.UsedPorts = defaultPorts
+
+	return resource, nil
+}
+
+func (e *EnvironmentScaler) calculateScaleDown(clusterName string, resourceProviders []*ResourceProvider) []*ResourceProvider {
+	var unusedProviders []*ResourceProvider
+
+	for _, provider := range resourceProviders {
+		if !provider.InUse {
+			unusedProviders = append(unusedProviders, provider)
+		}
+	}
+
+	return unusedProviders
+}
+
+func (e *EnvironmentScaler) calculateScaleUp(clusterName string, resourceProviders []*ResourceProvider, resourceConsumers []ResourceConsumer) ([]*ResourceProvider, []error) {
+	sorts := []struct {
+		Errs      []error
+		Providers []*ResourceProvider
+		SortBy    string
+	}{
+		{[]error{}, resourceProviders, "cpu"},
+		{[]error{}, resourceProviders, "mem"},
+	}
+
+	for _, s := range sorts {
+		switch s.SortBy {
+		case "cpu":
+			sortProvidersByCPU(s.Providers)
+			sortConsumersByCPU(resourceConsumers)
+
+		case "mem":
+			sortProvidersByMemory(s.Providers)
```

Comment:
why are we sorting twice using the same function each time? Shouldn't we have `SortByUsage` somewhere?

Reply:
@zpatrick Sorting by usage occurs after the `switch` statement, which is the only duplicated sorting function. The `switch` statement checks for whether a particular "run" of allocating consumers among providers is prioritizing CPU or memory, and sorts both the consumers and the providers accordingly. Then, the providers are sorted by usage to put the unused providers at the end of the list; since this should happen regardless of CPU vs memory priority, it happens after the `switch`.

Reply:
ah I see, I was reading too fast and thought it was 2 instances of `sortProvidersByMemory(s.Providers)` in a row. In regard to my comment above, it would be less LOC to use:
```go
sorters := []func([]*ResourceProvider, []ResourceConsumer){
    func(p []*ResourceProvider, c []ResourceConsumer) {
        sortProvidersMemory(p)
        sortProvidersByUsage(p)
        sortConsumersMemory(c)
    },
    func(p []*ResourceProvider, c []ResourceConsumer) {
        sortProvidersCPU(p)
        sortProvidersByUsage(p)
        sortConsumersCPU(c)
    },
}

for _, sort := range sorters {
    sort()

    for _, consumer := range resourceConsumers {
        ...
    }
}
```

Reply:
Worked in this suggestion. ✅